BIP 443: OP_CHECKCONTRACTVERIFY

bigspider opened the pull request:
Hi all,

This is a draft for the formal specifications of the `OP_CHECKCONTRACTVERIFY` (`CCV`) opcode.

`CCV` enables to build Script-based state machines that span across multiple transactions, by providing an ergonomic tool to commit to - and introspect - the Script and possibly some _data_ that is committed inside inputs or outputs.

Related to this PR:
- [Implementation in bitcoin-core](https://github.com/bitcoin/bitcoin/pull/32080)
- [Post on delving bitcoin](https://delvingbitcoin.org/t/op-checkcontractverify-and-its-amount-semantic/1527), focusing on the amount logic of the opcode.

Not covered in this draft:
- sigops budget (benchmarks needed)
- activation logic
- policy considerations (if any)

I recommend delving bitcoin for high level discussions about alternative implementations, applications, etc.

instagibbs commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+
+  # Amount checks
+
+  if flags == 0:
```
```suggestion
  if flags == CCV_FLAG_CHECK_OUTPUT:
```

bigspider replied:
Done in b2199950, thanks!

jirijakes commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+
+In short, the semantics of the opcode with respect to the Script can be summarized as follows:
+
+  Verify that the input/output with the given <code>index</code> is a P2TR output where the public key is obtained from <code><pk></code>, tweaked with the hash of <data> (if non-empty), then taptweaked with <code><taptree</code> (if present).
```
This [renders](https://github.com/Merkleize/bips/blob/ccv/bip-ccv.mediawiki#specification) as a code block without wrapping ⇒ have to scroll to the right to read it. Plain text (or blockquote, if you want to emphasize) might be better.

bigspider replied:
b2199950
I added the blockquote, and rephrased a bit - saying P2TR UTXO instead of the ambiguous 'output'. Does this make it clearer?

Sjors commented:
Some initial questions that come to mind:

1. Do you really need inputs to check other inputs? What does that enable? Otherwise my sense is that evaluating each input independently of others makes life easier (but you'll find out when implementing).

2. Can you make it so the output ordering doesn't matter?

3. You bring up "vector commitments" with no further comment, but it would be good to at least briefly explain what they're good for.

And see inline.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts be constrained with a signature, or with future introspection opcodes that allow fixing the amount.
```
It would be good to elaborate with an example.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+
+* UTXO sharing schemes like Ark, CoinPools, Timeout Trees, etc. use various types of output restrictions in order to enable multiple parties to share the control of a UTXO, while ensuring that each participant controls their own balance.
+* <code>OP_VAULT</code><ref>[[bip-0345.mediawiki|BIP-345]]</ref> is a proposed opcode to implement a 2-step withdrawal process, enabling on-chain reactive security.
+* <code>OP_CHECKTEMPLATEVERIFY</code><ref>[[bip-119.mediawiki|BIP-114]]</ref> is a long-proposed opcode to constrain the transaction to a ''template'' with a fixed set of outputs.
```
Ditto: having an example that (roughly) replicates `OP_CTV` would be useful

bigspider replied:
CCV alone doesn't replicate CTV, as it misses the 'exact amounts'. It could do it with `OP_AMOUNT` (although, of course, CTV is extremely efficient when you just want to commit to all the outputs - plus it has txid stability, if that's needed).

In this paragraph, my intention is to just set the context and hint that 'additional output restrictions are generally useful' - not compare proposals.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+* <code>OP_CHECKTEMPLATEVERIFY</code><ref>[[bip-119.mediawiki|BIP-114]]</ref> is a long-proposed opcode to constrain the transaction to a ''template'' with a fixed set of outputs.
+* Sidechains and rollups could be implemented via a UTXO encumbered with a recursive covenant, updating the sidechain state root every time it is spent.
+
+Constructions like BitVM<ref>https://bitvm.org/</ref> try to side-step the lack of a primitive allowing UTXOs to carry state with a clever use of Lamport Signatures, and optimistic execution of smart contracts. This comes with an extremely high cost in term of complexity, interactivity, and (potentially) in block size occupation, for some of the possible execution paths. Moreover, the design of f ...
```
Out of scope for a BIP, but maybe someone who works on BitVM can demonstrate the equivalent with `OP_CCV`? Presumably it doesn't completely replace it, but makes certain aspects more compact. Do they need additional op codes? Etc.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+
+This allows to embed a commitment to the data that can be validated during the Script execution, with an ad-hoc opcode called <code>OP_CHECKCONTRACTVERIFY</code>, while staying fully compatible with taproot. Notably:
+* the committed data does not make the UTXO any larger;
+* the keypath spend is still available to any party that possesses the private key of the naked key, as long as they have knowledge of the embedded data;
```
I'm confused by what you mean with "data", as opposed to a "program". Does this party have to know the pre-image or just the hash? And why is that helpful. Hopefully the future examples can illustrate this.

Sjors replied:
The implementation uses the term "double tweak". Perhaps that's a helpful term to explain the concept better?

bigspider replied:
> I'm confused by what you mean with "data", as opposed to a "program". Does this "party" have to know the pre-image or just the hash? And why is that helpful. Hopefully the future examples can illustrate this.

In b2199950 I added an additional introductory sentence in the paragraph, and also a footnote to explain how this could be used (which is anyway identical to taproot's keypath vs script spending, anyway; the point I'm trying to make is that this feature is not lost when using Scripts with CCV, despite _tampering_ with the internal key).


> And IIUC maybe use the term "single tweak" instead of "naked".

What I call the 'naked key' is the key before any tweak. So if you have no embedded data:
```
      naked_key ==(taptweak)==> taproot output key         (where naked_key == internal key)
```
Instead, if you have embedded data:
```
      naked_key ==(data tweak)==> internal_key ==(taptweak)==> taproot output key
```

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+== Examples ==
+
+This section documents some common Script fragments that use <code>OP_CHECKCONTRACTVERIFY</code> for various common choices of the parameters. Depending on the use case, some of the parameters might be passed via the witness stack.
+In these examples, <code><></code> (empty buffer) and <code>0</code> both refer to an empty stack element.
```
Can you frame these examples in the context of an actual application?

bigspider replied:
In 9bccf545930f9d4b8211f11d5bd93eac27b75695 I added one- or two-sentence explainers after each example (and deleted the example with just the _deduct_ mode, as the one right after already has a _deduct_ check followed by a _default_ one - which is the correct way of using it anyway).

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+The ability to constrain the future of coins beyond what is possible with presigned transactions is at the core of numerous attempts to improve bitcoin. Some of the proposed applications include:
+
+* UTXO sharing schemes like Ark, CoinPools, Timeout Trees, etc. use various types of output restrictions in order to enable multiple parties to share the control of a UTXO, while ensuring that each participant controls their own balance.
+* <code>OP_VAULT</code><ref>[[bip-0345.mediawiki|BIP-345]]</ref> is a proposed opcode to implement a 2-step withdrawal process, enabling on-chain reactive security.
```
Just saw this: https://github.com/bitcoin/bitcoin/blob/0a1d149112af2835166e3d6e62a925df2e416e4e/test/functional/feature_checkcontractverify_vaults.py

bigspider replied:
Yeah, I was a bit on the fence whether to put the code of the vault implementation in the BIP or in the bitcoin-core implementation, and ultimately I went for the latter, as functional tests are a lot more readable than dangling python scripts, or bare bitcoin Script fragments.

In general, I'm a bit concerned about focusing on the applications in the BIP, as that's extremely open-ended. I tried to focus on the _programming tool_ that CCV introduces. It will be mostly devs reading it, after all.

I feel the CCV-only vault is in a sweet spot as it's probably the smallest interesting construction that uses all the three modes (all except the _IGNORE_AMOUNT_ one)

Advise on the approach is welcome, though - please let me know if you would like a different approach, or some things to be covered more.

bigspider commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,416 @@
+* If the <code><index></code> is -1, it is replaced with the index of the current input.
+* If the <code><data></code> is the empty buffer, then there is no data tweak for the input/output being checked.
+
+Any other value of the parameters (except the <code><flags></code> as specified above) is invalid, and makes the opcode fail validation immediately.
```
Yes, undefined flags (which is more an extension of the opcode) are left for possible future extensions, while that seems dangerous/footgun-prone for the other parameters (which could be passed via the witness)
In b2199950 I tried to make it more explicit by explicitly listing all the parameters.

bigspider commented:
> 1. Do you really need inputs to check other inputs? What does that enable? Otherwise my sense is that evaluating each input independently of others makes life easier (but you'll find out when implementing).

I don't think it would make it easier; in fact, apart from the amount semantic (amounts are not checked when you CCV-check and input), the opcode's behavior is currently very symmetric. So in a way, I think the current formulation is simpler by not restricting to just the current input (while staying more general).

In everything I worked on so far (even in combinations with other opcodes, e.g. see [pymatt](https://github.com/Merkleize/pymatt)), checking just the current input is indeed sufficient. However, I think it's likely that it would be useful in some more advanced constructions. For example, you could create a separate _sentinel_ UTXO, and have some spending condition (possibly for many UTXOs at the same time) be only available if the sentinel UTXO is present in the transaction. Once the sentinel is spent, those spending conditions become unavailable for all the remaining. That seems similar in spirit to the _connector outputs_  used in Ark, but without presigned transactions, so I suspect it's useful.

> 2. Can you make it so the output ordering doesn't matter?

You can leave flexibility as to where outputs must be by either
- using -1 for the output index (meaning, the corresponding output must be in the same position as the input index); this would for example allow batching multiple 1-input-1-output CCV-encumbered spends (where each input must "produce" its own output, without aggregation) in the same transaction
- passing the `<index>` parameter in the witness of the transaction. E.g. in the [vault example](https://github.com/bitcoin/bitcoin/pull/32080/files#diff-e796b699bc9cf1f5b7f2eb10c2d88f5c47b7701d5835ca7734fbfafd7d72e87c), the _revault index_ and the _trigger index_ of the `trigger_and_revault` clause are arbitrary in this way.

> 3. You bring up "vector commitments" with no further comment, but it would be good to at least briefly explain what they're good for.

I'll add a footnote, thanks!

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+In the following, the ''current input'' is the input whose Script is being executed.
+
+The following value of the <code><flags></code> are defined:
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
```
Most flags in Bitcoin Core are usigned integers, which seems easier to read. But I guess this is because the script interpreter uses (variable size) signed integers? On the bright side, it means you can expand the number of bit flags easily...

When this flag is absent, I assume you don't check the input?

bigspider replied:
The flag can't be 'absent', as it's the value of the `flags` parameter. Note that it's not a bitmap: -1 means "check input", 0, 1 and 2 are the three defferent _amount_ behaviors for checking the output (default, deduct and ignore).

Perhaps the confusion comes from calling it `flags` - let me know if you have suggestions for a better name.

Sjors replied:
At least to me flags imply a bitmap. Terminology aside, why not use a bitmap?

Since they're all mututally exclusive, I would probably call it "mode".

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+
+The following value of the <code><flags></code> are defined:
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
+* <code>CCV_FLAG_CHECK_OUTPUT = 0</code>: Check an output's script; preserve the (possibly residual) amount.
```
When this flag is absent, you don't check the output?

And `CCV_FLAG_CHECK_INPUT` can be used in any combination with `CCV_FLAG_CHECK_INPUT`, including not checking either?

bigspider replied:
As per [this comment](https://github.com/bitcoin/bips/pull/1793#discussion_r2013047590), the defined flags are all mutually exclusive.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
+* <code>CCV_FLAG_CHECK_OUTPUT = 0</code>: Check an output's script; preserve the (possibly residual) amount.
+* <code>CCV_FLAG_IGNORE_OUTPUT_AMOUNT = 1</code>: Check an output's script; ignore amounts.
+* <code>CCV_FLAG_DEDUCT_OUTPUT_AMOUNT = 2</code>: Check an output's script; deduct the output amount from the input's residual amount.
```
For these two I assume `CCV_FLAG_CHECK_OUTPUT` must be set? And they can't be both set.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+Any other value of the <code><flags></code> makes the opcode succeed validation immediately for the current input<ref>This allows to soft-fork future behavior by introducing new values for the <code><flags></code>. As the flags would always be hard-coded via a push in the Script, the risk of mistakes seems negligible.</ref>.
+
+The following values of the other parameters have special meanings:
+* If the <code><taptree></code> is -1, it is replaced with the Merkle root of the current input's tapscript tree. If the taptree is the empty buffer, then the taptweak is skipped.
```
Have you considered using flags for these special cases and then having fewer stack elements? That seems both more space efficient and probably makes scripts easier to read.

bigspider replied:
All the flags and the special parameters are encoded as 1-byte push opcodes, so I don't think you can really save much by aggregating them in the flags (e.g. using extra bits that are currently in the `OP_SUCCESS` behavior).
Moreover, I think it would be very inconvenient (both in implementation and in Script readability) if the number of stack arguments is different for different use cases of CCV.

Sjors commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+<code>BIP341_NUMS_KEY</code> is the x-only provably unspendable key <code>50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in [[bip-341.mediawiki|BIP-341]].
+
+<source lang="python">
+  if flags < CCV_FLAG_CHECK_INPUT or flags > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
```
I assume you mean to check that no unknown flags are in use? This seems wrong though, because `CCV_FLAG_DEDUCT_OUTPUT_AMOUNT | CCV_FLAG_IGNORE_OUTPUT_AMOUNT > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT`

Probably more clear to have something like `ALL_FLAGS = ... | CCV_FLAG_DEDUCT_OUTPUT_AMOUNT | CCV_FLAG_IGNORE_OUTPUT_AMOUNT | ...` and check `flags && !ALL_FLAGS == 0`

bigspider replied:
Since flags are mutually exclusive and the defined values are -1, 0, 1, 2, this is just checking if you're out of the range (then, _SUCCESS_ as an upgrade hook to add future flags).

Sjors replied:
See above comment, I think calling it "mode" would make this code clear:

```
if mode < CCV_FLAG_CHECK_INPUT or mode > CCV_FLAG_DEDUCT_OUTPUT_AMOUNT:
```

bigspider replied:
Changed to `mode` instead of `flags` in 4ed71843e51742b6a60f445c272b73a1e90dc10f.

Sjors commented:
Do you have any thoughts on what this could like in terms of descriptors? Either in the general case of anything that can be expressed in miniscript, or more narrowly for something like a vault?

E.g. fresh vault deposit addresses could be generated from something like `tr(cold,{trigger_leaf})`, where `trigger_leaf` could be something like `ccv(-1,{withdraw_leaf,recover_leaf},0,...)`, etc.

Since programs can take arbitrary data, e.g. a withdrawal address for a vault, I could imagine that when a wallet signs (or detects an unauthorized) a trigger transaction it generates a descriptor for that, so that it knows how to spend it (and when).

But how generalisable is that?

Sjors commented:
>> Do you really need inputs to check other inputs? What does that enable?

> However, I think it's likely that it would be useful in some more advanced constructions. For example, you could create a separate sentinel UTXO,

Just saw the mailinglist thread about this from 2023: https://gnusha.org/pi/bitcoindev/CALZpt+F251k7gSpogwFYHxFtGxc_tZjB4UU4SVEr=WvrsyMVMQ@mail.gmail.com/

The BIP should probably address the pros and cons of "cross-input inspection".

bigspider commented:
> Do you have any thoughts on what this could look like in terms of descriptors? Either in the general case of anything that can be expressed in miniscript, or more narrowly for something like a vault?

@sanket1729 has given several great talks on generalizing miniscript for covenant use cases, you can look for those if you're interested.

My general take is that descriptors are the wrong tool for this purpose: a spend from UTXO X to UTXO Y where f(Y) = X needs to somehow encode the relation between X and Y as a _predicate_. While you could do that (every program is a predicate, and every program expressible in Script is a predicate that can be expressed in a tree structure like miniscript...) it quickly becomes unmanageable.

For CCV, what works well (kinda by design) is to think in terms of _states_ and _state transitions_. You can check [these docs](https://github.com/Merkleize/pymatt/blob/master/docs/contracts.md) and the code examples in the [pymatt](https://github.com/Merkleize/pymatt) repo if you're interested in more details in how I'm framing it - the PR in bitcoin-core strips most of those useful abstractions for the sake of conciseness, but that's certainly not how one would write those contracts in practice.

The framework has some nice properties: once you define the contracts, all you need to know is the contract definition, and the initial parameters. Everything else can be deterministically derived from that, and the blockchain. You can scan for the initial UTXOs matching that contract, and given a transaction spending those UTXOs, you can deterministically deduce what are the next states, parameters (and data, if any) of the new UTXOs that are produced. That's even when someone else in the contract made transactions, as it's easy to 'decode' the witness to understand what _clause_ (tapleaf) was used, and with what arguments. 

> Just saw the mailinglist thread about this from 2023: https://gnusha.org/pi/bitcoindev/CALZpt+F251k7gSpogwFYHxFtGxc_tZjB4UU4SVEr=WvrsyMVMQ@mail.gmail.com/
> 
> The BIP should probably address the pros and cons of "cross-input inspection".

I personally think speculation on all the possible ways people might use (and abuse) the opcode is out of scope, and it quickly gets unmanageable - because there is an infinite number of ways of (per-)using the opcode. For example, both [`OP_CAT`](https://github.com/bitcoin/bips/blob/master/bip-0347.mediawiki) and [`OP_TXHASH`](https://github.com/bitcoin/bips/pull/1500) both enable fine-grained introspection of other inputs, and would therefore raise the same concern - if one thinks that is a valid concern; yet the BIPs don't discuss this.

IMHO that kind of discussion would be more fitting for a BIP describing a proposal to activate a certain package of opcodes as a soft-fork. OP_CCV is not meant to be activated as a stand-alone opcode (that would be kinda silly for a number of reasons). Such a BIP could in fact focus on the _capabilities_, ignoring the exact details of the opcodes (that are not interesting in that context).

Sjors commented:
> @sanket1729 has given several great talks on generalizing miniscript for covenant use cases, you can look for those if you're interested.

Found one, which I'll watch: https://www.youtube.com/watch?v=xNAn9LTzk2g

> My general take is that descriptors are the wrong tool for this purpose:

> You can check [these docs](https://github.com/Merkleize/pymatt/blob/master/docs/contracts.md) and the code examples in the [pymatt](https://github.com/Merkleize/pymatt) repo if you're interested in more details in how I'm framing it

I'll look into that. With previous soft forks like SegWit and Taproot it took many years after activation for enough tooling to be developed to fully take advantage. I'm trying to get a sense of that in this case. It seems fine keep the actual Bitcoin Core implementation simpler.

> IMHO that kind of discussion would be more fitting for a BIP describing a proposal to activate a certain package of opcodes as a soft-fork

That makes sense. My impression so far is that there exists some sort of covenant threshold beyond which we open the pandora's box, and it doesn't really matter how we cross the threshold.

Perhaps there could be a BIP that discusses several combinations of op codes, explain which combinations cross the threshold, and then explain why that's fine, or not.

If it's fine to cross the threshold, I tend to think we should go for the most powerful tool(s) with simplest and safest implementation. 

If it's not fine, we would be limited to extremely restricted op codes that give us desired functionality without crossing the threshold.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,420 @@
+</pre>
+
+
+== Introduction ==
```
Nit: Missing "Abstract" section.

bigspider replied:
9bccf545930f9d4b8211f11d5bd93eac27b75695: I renamed the `Introduction` to `Abstract`.

murchandamus reviewed (commented):
The formatting looks mostly good, got one nit. This seems to have some action already, so I will take a look at the content when the open feedback has been processed.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+This opcode enables users to create UTXOs that carry a dynamic commitment to a piece of data. The commitment can be validated during the execution of the Script, allowing introspection to the committed data. Moreover, a Script can constrain the program (internal public key and taptree) and the data of one or more outputs.
+
+In conjunction with an opcode for ''vector commitments''<ref>''Vector commitments'' are cryptographic primitives that allow to commit to a vector of values via single short value. Hashing and concatenation trivially allow to commit to an entire vector, and later reveal all of its elements. Merkle trees are among the simplest efficient vector commitments, allowing to reveal individual elements wit ...
```
It would be easier to provide detailed suggestions if the lines were broken to a shorter length, e.g., 120 characters, instead of whole paragraphs in one line.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+* UTXO sharing schemes like Ark, CoinPools, Timeout Trees, etc. use various types of output restrictions in order to enable multiple parties to share the control of a UTXO, while ensuring that each participant controls their own balance.
+* <code>OP_VAULT</code><ref>[[bip-0345.mediawiki|BIP-345]]</ref> is a proposed opcode to implement a 2-step withdrawal process, enabling on-chain reactive security.
+* <code>OP_CHECKTEMPLATEVERIFY</code><ref>[[bip-119.mediawiki|BIP-114]]</ref> is a long-proposed opcode to constrain the transaction to a ''template'' with a fixed set of outputs.
```
```suggestion
* <code>OP_CHECKTEMPLATEVERIFY</code><ref>[[bip-119.mediawiki|BIP-114]]</ref> is a long-proposed opcode to constrain a transaction to a ''template'' with a fixed set of outputs.
```

"The transaction" begs the question "which transaction", but I think you simply mean "a transaction".

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+== Design ==
+
+<code>OP_CHECKCONTRACTVERIFY</code> is an implementation of a new primitive that could be called ''state-carrying UTXOs''. It allows to embed a commitment to a piece of data in a UTXO, and to validate it during the execution of the Script, and ''carry'' a (possibly dyamically computed) piece of data to the new UTXOs that are produced.
```
re: Capitalized "Script" see above.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+<code>OP_CHECKCONTRACTVERIFY</code> is an implementation of a new primitive that could be called ''state-carrying UTXOs''. It allows to embed a commitment to a piece of data in a UTXO, and to validate it during the execution of the Script, and ''carry'' a (possibly dyamically computed) piece of data to the new UTXOs that are produced.
+
+We consider the ''program'' of a P2TR UTXO to be composed of an x-only public key (that we call the ''naked key''), and a Merkle tree of Scripts. If there is no data committed in the UTXO, then the naked key is the internal key as defined in BIP-341.
```
It seems to me that this is the definition for "program" that I was missing above. I would suggest either moving the definition to the first use, or sidestepping the use of the term above.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+This allows to embed a commitment to the data that can be validated during the Script execution, while staying fully compatible with taproot. Notably:
+* the committed data does not make the UTXO any larger;
+* the keypath spend is still available to any party that possesses the private key of the naked key, as long as they have knowledge of the embedded data (or at least their hash)<ref>For example, in a multi-party contract, the naked key could be aggregate key using [[bip-0327.mediawiki|MuSig2]]; the taproot keypath would therefore allow a ''cooperative'' spend, without executing any Script at all. ...
```
```suggestion
* the keypath spend is still available to any party that possesses the private key of the naked key, as long as they have knowledge of the embedded data (or at least the data’s hash)<ref>For example, in a multi-party contract, the naked key could be aggregate key using [[bip-0327.mediawiki|MuSig2]]; the taproot keypath would therefore allow a ''cooperative'' spend, without executing any Script at all. Like for all taproot transactions, this is indeed the cheapest way of spending the UTXO - albeit not always possible in practice.</ref>;
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+This allows to embed a commitment to the data that can be validated during the Script execution, while staying fully compatible with taproot. Notably:
+* the committed data does not make the UTXO any larger;
+* the keypath spend is still available to any party that possesses the private key of the naked key, as long as they have knowledge of the embedded data (or at least their hash)<ref>For example, in a multi-party contract, the naked key could be aggregate key using [[bip-0327.mediawiki|MuSig2]]; the taproot keypath would therefore allow a ''cooperative'' spend, without executing any Script at all. ...
+* if multiple Scripts are in different leaves of the taptree, only the ones that actually need to access the data have to pay a cost for it, in terms of additional witness bytes.
```
Is it possible that "Script" always refers to a leaf script?

```suggestion
* if multiple Scripts are in different leaves of the taptree, only the ones that actually need to access the data have to pay a cost for it, in terms of additional witness bytes.
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+=== Output amounts ===
+
+When checking the Script of one or more output with <code>OP_CHECKCONTRACTVERIFY</code>, it is usually necessary to also check that the amount of the current UTXO is correctly distributed among the outputs in the expected way. Therefore, the opcode already includes an amount semantic that covers the common use cases.
```
What is "the current UTXO"? Is this referring to an input that spends an OP_CHECKCONTRACTVERIFY-encumbered TXO?

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+When checking the Script of one or more output with <code>OP_CHECKCONTRACTVERIFY</code>, it is usually necessary to also check that the amount of the current UTXO is correctly distributed among the outputs in the expected way. Therefore, the opcode already includes an amount semantic that covers the common use cases.
+
+There are three supported modes for the opcode when checking an output, depending on the value of the <code>mode</code> parameter:
```
Where does the opcode that is being checked stand? Is it in the output itself, or are we talking about a previous encumbrance?

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+When checking the Script of one or more output with <code>OP_CHECKCONTRACTVERIFY</code>, it is usually necessary to also check that the amount of the current UTXO is correctly distributed among the outputs in the expected way. Therefore, the opcode already includes an amount semantic that covers the common use cases.
+
+There are three supported modes for the opcode when checking an output, depending on the value of the <code>mode</code> parameter:
+* ''default'': the residual amount of the current input must be preserved in the output (aggregate across the inputs that specify the output);
```
Does current input refer to the input at the same index as this output or what makes it "current"? What happens when there are multiple inputs spending TXOs that were encumbered with `OP_CHECKCONTRACTVERIFY`?

```suggestion
* ''default'': the residual amount of the current input must be preserved in the output (aggregate across the inputs that specify the output);
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+* ''ignore'': the output amount is ignored.
+* ''deduct'': the amount of the checked output is subtracted from the amount of the current input (the residual amount is then available for further checks);
+
+The ''default'' logic covers the common case where an input wishes to send the full amount to a specific output.
```
This seems to assign a strange sense of agency to the input. Do I understand that you mean something like:

```suggestion
The ''default'' logic covers the common case where a UTXO’s full amount is required to be sent to a specific output.
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+The ''default'' logic covers the common case where an input wishes to send the full amount to a specific output.
+
+The ''deduct'' logic allows to assign portions of the input amount to one or more outputs; the residual amount, checked with a final check using the ''default'' logic, can be used to enforce that the total amount is preserved.
```
If there are multiple UTXOs, what happens with the change? If there is only one UTXO, how do you spend the entire amount without bringing more funds to pay for fees?

bigspider replied:
What is 'change'? The amount has to be entirely assigned to the outputs.
For applications like vaults, you can do use the _deduct_ logic with an output that has the same script as the input; most contracts are non-recursive, so they always 'move forward', and there is no meaningful definition of change.

Fees have to be paid exogenously, with a separate input (or by spending an output, if that applies). I think this is necessary in any kind of covenant, as the fee is by definition an amount that is not encumbered by the covenant.

murchandamus replied:
I may have not realized then yet that the fees are always paid exogenously.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+The following figures illustrate some common examples of supported use cases for the amount logic. This list is not exhaustive, as there are many more possible combinations.
+
+'''Remark:''' validation fails if the amount of an output is checked with both the ''default'' and the ''deduct'' logic in the same transaction, or multiple times with the ''deduct'' logic. This prevents duplicate or inconsistent counting of the same amounts.
+
+'''Remark:''' it is allowed to check for multiple inputs to check the same output with the ''default'' logic. This allows multiple inputs to aggregate (in full or in part) their amounts to the same output.
```
I must have failed to pick up on some part of the logic above, but how are contracts tied to specific outputs? What happens if you have two UTXOs tied to the same contract that requires an equal split of the funds between A and B? Let’s say 50% is deducted to A and then the rest is defaulted to B. Would you need two outputs to A and one to B to achieve this?

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts should be constrained either with a signature, or with a future introspection opcodes that allow fixing the amount. In lack of that, amounts would be malleable.
```
```suggestion
Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts should be constrained either with a signature, or with future introspection opcodes that allow fixing the amount. In lack of that, amounts would be malleable.
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+== Specification ==
+
+The tapscript opcode <code>OP_SUCCESS187</code> (<code>0xbb</code>) is constrained with new rules to implement <code>OP_CHECKCONTRACTVERIFY</code>.
```
Note that [BIP 345: OP_VAULT](https://github.com/bitcoin/bips/blob/3ccc59dbdbca13fa5cc4ebeba9180946fa122003/bip-0345.mediawiki#L225) also proposes to use OP_SUCCESS187.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+When evaluating <code>OP_CHECKCONTRACTVERIFY</code> (<code>OP_SUCCESS187</code>,
+<code>0xbb</code>), the expected format of the stack, shown top to bottom, is:
+
+<source>
+<mode>
+<taptree>
+<pk>
+<index>
+<data>
+</source>
```
AFAIA, the content of the [stack](https://en.bitcoin.it/wiki/Script#Stack) is usually written in one line bottom to top:

```suggestion
<source>
<data> <index> <pk> <taptree> <mode> 
</source>
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+  data_tweak = sha256(pk || data)
+</source>
+
+In the following, the ''current input'' is the input whose Script is being executed.
```
"Current input" was used above without an introduction of the term. Please introduce the term on first use.

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+The following values of the other parameters have special meanings:
+* If the <code><taptree></code> is -1, it is replaced with the Merkle root of the current input's tapscript tree. If the taptree is the empty buffer, then the taptweak is skipped.
+* If the <code><pk></code> is 0, it is replaced with it is replaced with the NUMS x-only pubkey <code>0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in BIP-0340. If the <code><pk></code> is -1, it is replaced with the taproot internal key of the current input.
```
```suggestion
* If the <code><pk></code> is 0, it is replaced with the NUMS x-only pubkey <code>0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in BIP-0340. If the <code><pk></code> is -1, it is replaced with the taproot internal key of the current input.
```

murchandamus commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,413 @@
+
+=== Script support for <code>OP_CHECKCONTRACTVERIFY</code> ===
+
+The specification is divided in three parts:
```
```suggestion
The specification is divided into three parts:
```

murchandamus reviewed (commented):
I only got up to the start of the section "Script support for <code>OP_CHECKCONTRACTVERIFY</code>" this time.

Christewart commented on bip-ccv.mediawiki:
```diff
@@ -0,0 +1,409 @@
+The following values of the other parameters have special meanings:
+* If the <code><taptree></code> is -1, it is replaced with the Merkle root of the current input's tapscript tree. If the taptree is the empty buffer, then the taptweak is skipped.
+* If the <code><pk></code> is 0, it is replaced with the NUMS x-only pubkey <code>0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0</code> defined in BIP-0340. If the <code><pk></code> is -1, it is replaced with the taproot internal key of the current input.
+* If the <code><index></code> is -1, it is replaced with the index of the current input.
```
I don't like the usage of `-1` as default arguments for several reasons

1. It opens up malleability vectors. IIUC this proposal correctly, we have 3 different usages of `-1` on the witness stack. This means we have 8 different combinations of potential valid values on the witness stack. I've pasted a table below of the different malleabilit combinations
2. Its not composable with other index based opcodes. Other index based opcodes would need to adopt your `-1` semantic to be composable with `OP_CCV`. 

Combination # | Taptree | PK | IDX
-- | -- | -- | --
1 | taptree | pk | idx
2 | taptree | pk | -1
3 | taptree | -1 | idx
4 | taptree | -1 | -1
5 | -1 | pk | idx
6 | -1 | pk | -1
7 | -1 | -1 | idx
8 | -1 | -1 | -1

If you remove the `-1` semantic you would get rid of malleability potential and make OP_CCV more composable with future index based opcodes ([`OP_INOUT_AMOUNT`](https://delvingbitcoin.org/t/op-inout-amount/549/4), and [various elements project opcodes](https://github.com/ElementsProject/elements/blob/7ed597f4a5f713e33aac04c87c1a9da5ecd7d52c/src/script/interpreter.cpp#L1925))

bigspider replied:
> 1. It opens up malleability vectors. IIUC this proposal correctly, we have 3 different usages of `-1` on the witness stack. This means we have 8 different combinations of potential valid values on the witness stack. I've pasted a table below of the different malleability combinations

In the vast majority of usages, all those arguments would not be passed via the witness stack. The only case where I used an index passed via the witness stack, so far, was for `index` argument in vaults, in order to make them more ergonomic (in that you have more freedom in how to compose your transactions).

If a script takes those arguments from the witness, then it's up to the script designer to deal with malleability, possibly by adding additional constraints (like a `CHECKSIG`) in order to avoid it.

**Remark**: I don't it would ever make sense for `pk` or `taptree` to directly come from the witness; in some case, it could come from a computation that depends on the witness arguments - for example, as the result of opening a commitment, which implicitly deals with malleability,

> 2. Its not composable with other index based opcodes. Other index based opcodes would need to adopt your `-1` semantic to be composable with `OP_CCV`.

Do you have a concrete example of lack of composability?

> If you remove the `-1` semantic you would get rid of malleability potential and make OP_CCV more composable with future index based opcodes ([`OP_INOUT_AMOUNT`](https://delvingbitcoin.org/t/op-inout-amount/549/4), and [various elements project opcodes](https://github.com/ElementsProject/elements/blob/7ed597f4a5f713e33aac04c87c1a9da5ecd7d52c/src/script/interpreter.cpp#L1925))
> This has the tradeoff of increasing witness size, but I think the tradeoff is worth it for the benefit of composability. These values can be useful for other opcodes other than `OP_CCV` during script validation! :-)

The `-1` semantic is extremely common and dramatically optimizes the script sizes.

The `CCV_MODE_CHECK_OUTPUT_IGNORE_AMOUNT` mode is there precisely to future-proof the opcode and make it composable with any other amount introspection - even if I don't currently have a use case for it.

It's unclear to me how removing a feature could make the opcode more composable. If you don't need the feature for your script, you can simply not use it.

Christewart replied:
I'll reply on delving or the mailing list as this probably isn't the best place for posting long form results and to start the conversation about potentially use design changes. Apologies for starting the conversation here and not a more appropriate venue. I should have something ready in the next few days.

murchandamus reviewed (commented):
Not a full review, more of a quick check of what has happened here lately and a skim of the changes based on my prior review.

I see a few todos in the document, could you describe in a couple sentences how you see the status of this document?

murchandamus commented:
Let’s refer to this proposal as **BIP 443**. Could you please rename the file, add an entry to the table, and update the number and created date in the preamble?

bigspider commented:
Updated header per number assignment, and split the long lines in bec994f70c48de2c0f5b475c6d6c32e55a242fd0; however, that breaks for lists, so I partially undid it c950a1e4fcbb5a415dfdc3cd18ed9963d590e1fc.

@murchandamus would it be a good time to squash all commits? I think there are no unaddressed review comments at this time.

murchandamus commented:
> @murchandamus would it be a good time to squash all commits? I think there are no unaddressed review comments at this time.

Sure that sounds reasonable

bigspider commented:
Squashed all commits and rebased.

murchandamus reviewed (commented):
As discussed, this will still need to be amended in a few details to be ready for Proposed, but looks ready to be merged as Draft.

Sjors commented:
Can you expand the BIP to mention that fees have to paid exogenously?https://github.com/bitcoin/bips/pull/1793#discussion_r2045289706

That said, a mode similar to `default` that allows any amount to be missing as long as it goes to fees, could be useful.

jonatack commented:
> Can you expand the BIP to mention that fees have to be paid exogenously?[#1793 (comment)](https://github.com/bitcoin/bips/pull/1793#discussion_r2045289706)
> 
> That said, a mode similar to `default` that allows any amount to be missing as long as it goes to fees, could be useful.
> 
> In particular there's a privacy advantage when you don't have to use coins from some other wallet to initiate and/or cancel a vault withdrawal.
> 
> The trade-off there is that an attacker can (threaten to) burn your wallet, but they can't take it.

@bigspider would you like to follow up on @Sjors' feedback?

jonatack commented on bip-0443.mediawiki:
```diff
@@ -0,0 +1,465 @@
+The ability to constrain the future of coins beyond what is possible with presigned transactions is at the core of
+numerous attempts to improve bitcoin. Some of the proposed applications include:
+
+* UTXO sharing schemes like Ark, CoinPools, Timeout Trees, etc. use various types of output restrictions in order to enable multiple parties to share the control of a UTXO, while ensuring that each participant controls their own `balance.
```
s/`balance/balance/ ... done in #1864

bigspider commented:
> Can you expand the BIP to mention that fees have to be paid exogenously?[#1793 (comment)](https://github.com/bitcoin/bips/pull/1793#discussion_r2045289706)

I can add a comment about it if you think it's useful. However, I believe this is not specific to CCV: for any covenant that constrains an output Script, it would be highly unusual to not constrain amounts, as sending 0 to an output is equivalent to not having a covenant at all. [CTV](https://github.com/bitcoin/bips/blob/master/bip-0119.mediawiki) and [VAULT](https://github.com/bitcoin/bips/blob/master/bip-0345.mediawiki) reached the same conclusion.

> That said, a mode similar to `default` that allows any amount to be missing as long as it goes to fees, could be useful.
> 
> In particular there's a privacy advantage when you don't have to use coins from some other wallet to initiate and/or cancel a vault withdrawal.
> 
> The trade-off there is that an attacker can (threaten to) burn your wallet, but they can't take it.

If the attacker is a miner, they can take the money. I'm not sure it would be a good idea to enshrine in the opcode a semantic that is inherently vulnerable to miner's theft.

Besides, I don't see a way of defining a semantic that is not ambiguous, once you take other inputs into account: it wouldn't be unequivocally determined _whose_ amount is 'going' to fees, and whose is 'going' to the corresponding output.

Sjors commented:
> If the attacker is a miner, they can take the money. I'm not sure it would be a good idea to enshrine in the opcode a semantic that is inherently vulnerable to miner's theft.

Or colludes with a miner. One could even imagine a malevolent accelerator that pays out its users half the fee for any transaction wasn't already in their mempool.

I do think it's good to clarify this, since not everyone reading this BIP will have read CTV and VAULT.
